File: python/백준/그래프/DFS/2644 촌수계산.py
import sys
sys.setrecursionlimit(10**6)

input = sys.stdin.readline

n = int(input())
start, end = map(int, input().split())

graph = [[] for _ in range(n+1)]

m = int(input())

for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

visited = [-1] * (n + 1)

def dfs(node, dist):
    if node == end:
        return dist
    
    for next_node in graph[node]:
        if visited[next_node] == -1:
            visited[next_node] = dist + 1
            result = dfs(next_node, dist + 1)
            if result != -1:
                return result
    
    return -1

visited[start] = 0
print(dfs(start, 0))

File: cpp/백준/그래프/DFS/2644 촌수계산.cpp
#include <iostream>
#include <vector>
using namespace std;

int n, start, endNode;
vector<vector<int>> graph;
vector<int> visited;

int dfs(int node, int dist) {
    if (node == endNode) return dist;
    for (int next_node : graph[node]) {
        if (visited[next_node] == -1) {
            visited[next_node] = dist + 1;
            int result = dfs(next_node, dist + 1);
            if (result != -1) return result;
        }
    }
    return -1;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n;
    cin >> start >> endNode;

    graph.assign(n + 1, {});

    int m;
    cin >> m;

    for (int i = 0; i < m; ++i) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    visited.assign(n + 1, -1);
    visited[start] = 0;
    cout << dfs(start, 0) << "\n";
    return 0;
}
